pkg/query/selection.go

```diff
@@ -0,0 +1,299 @@
+func (op Operator) IsNullable() bool {
+	return op == EqualsOrNilOperator
+}
+
+// IsNumeric returns true if the operator works only with numeric operands
+func (op Operator) IsNumeric() bool {
+	return op == LessThanOperator || op == GreaterThanOperator
+}
+
+var operators = []Operator{EqualsOperator, NotEqualsOperator, InOperator,
+	NotInOperator, GreaterThanOperator, LessThanOperator, EqualsOrNilOperator}
+
+const (
+	// OpenBracket is the token that denotes the beginning of a multivariate operand
+	OpenBracket string = "["
+	// OpenBracket is the token that denotes the end of a multivariate operand
+	CloseBracket string = "]"
+	// Separator is the separator between field and label queries
+	Separator        rune   = ','
+	OperandSeparator string = "+"
+)
+
+// CriterionType is a type of criteria to be applied when querying
+type CriterionType string
+
+const (
+	// FieldQuery denotes that the query should be executed on the entity's fields
+	FieldQuery CriterionType = "fieldQuery"
+	// LabelQuery denotes that the query should be executed on the entity's labels
+	LabelQuery CriterionType = "labelQuery"
+)
+
+var supportedQueryTypes = []CriterionType{FieldQuery, LabelQuery}
+
+// UnsupportedQueryError is an error to show that the provided query cannot be executed
+type UnsupportedQueryError struct {
+	Message string
+}
+
+func (uq *UnsupportedQueryError) Error() string {
+	return uq.Message
+}
+
+// Criterion is a single part of a query criteria
+type Criterion struct {
+	// LeftOp is the left operand in the query
+	LeftOp string
+	// Operator is the query operator
+	Operator Operator
+	// RightOp is the right operand in the query which can be multivariate
+	RightOp []string
+	// Type is the type of the query
+	Type CriterionType
+}
+
+// ByField constructs a new criterion for field querying
+func ByField(operator Operator, leftOp string, rightOp ...string) Criterion {
+	return newCriterion(leftOp, operator, rightOp, FieldQuery)
+}
+
+// ByLabel constructs a new criterion for label querying
+func ByLabel(operator Operator, leftOp string, rightOp ...string) Criterion {
+	return newCriterion(leftOp, operator, rightOp, LabelQuery)
+}
+
+func newCriterion(leftOp string, operator Operator, rightOp []string, criteriaType CriterionType) Criterion {
+	return Criterion{LeftOp: leftOp, Operator: operator, RightOp: rightOp, Type: criteriaType}
+}
+
+func (c Criterion) Validate() error {
+	if len(c.RightOp) > 1 && !c.Operator.IsMultiVariate() {
+		return fmt.Errorf("multiple values received for single value operation")
+	}
+	if c.Operator.IsNullable() && c.Type != FieldQuery {
+		return &UnsupportedQueryError{"nullable operations are supported only for field queries"}
+	}
+	if c.Operator.IsNumeric() && !isNumeric(c.RightOp[0]) {
+		return &UnsupportedQueryError{Message: fmt.Sprintf("%s is numeric operator, but the right operand is not numeric", c.Operator)}
+	}
+	//if slice.StringsAnyEquals(c.RightOp, "") {
```

Comment:
clean up or uncomment, also add a test for the case  (per op?) in case one is missing